File: api/app/services/event_service.py
from datetime import datetime, timedelta
from typing import List, Optional
from fastapi import HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, func
import logging
from sqlalchemy.types import String, Interval, Time

from app.models.event import Event
from app.schemas.event import EventCreate, EventResponse

# Configure logging
logging.basicConfig(
    level=logging.DEBUG, 
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

class EventService:
    def __init__(self, db: Session):
        self.db = db

    def create_event(self, event: EventCreate) -> Event:
        """Create a new event with proper validation and error handling"""
        logger.info(f"Creating event: {event}")
        
        try:
            # Ensure start_time is UTC
            if isinstance(event.start_time, str):
                event.start_time = datetime.fromisoformat(event.start_time.replace('Z', '+00:00'))
            
            # ENHANCED LOGGING FOR DURATION
            logger.critical(f"DURATION VERIFICATION - Raw Duration: {event.duration}")
            logger.critical(f"DURATION VERIFICATION - Duration Type: {type(event.duration)}")
            
            # Log pre-conflict check details
            logger.debug(f"Pre-conflict check event details:")
            logger.debug(f"Start Time: {event.start_time}")
            logger.debug(f"Duration: {event.duration} minutes")
            logger.debug(f"Is Recurring: {event.is_recurring}")
            logger.debug(f"Recurring Days: {event.recurring_days}")
            
            # Check for conflicts BEFORE creating the event
            if conflict := self._get_conflict_details(event):
                logger.warning(f"Event creation blocked due to conflict: {conflict}")
                raise HTTPException(status_code=400, detail=conflict)
            
            # Convert event to dict and handle recurring days
            event_dict = event.model_dump()
            
            # Convert recurring_days to comma-separated string if it's a list
            if event_dict.get('recurring_days'):
                event_dict['recurring_days'] = ','.join(sorted(event_dict['recurring_days']))
            
            # Log final event details before database insertion
            logger.debug(f"Final event details for database insertion: {event_dict}")
            
            # Create and save event
            db_event = Event(**event_dict)
            self.db.add(db_event)
            self.db.flush()  # Flush to get the ID without committing
            
            # Commit the transaction
            self.db.commit()
            self.db.refresh(db_event)
            
            # Detailed logging of created event
            logger.info(f"Event created successfully:")
            logger.info(f"Event ID: {db_event.id}")
            logger.info(f"Event Name: {db_event.name}")
            logger.info(f"Event Start Time: {db_event.start_time}")
            logger.info(f"Event Duration: {db_event.duration} minutes")
            logger.info(f"Is Recurring: {db_event.is_recurring}")
            logger.info(f"Recurring Days: {db_event.recurring_days}")
            
            return db_event
            
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Unexpected error creating event: {str(e)}", exc_info=True)
            self.db.rollback()
            raise HTTPException(status_code=500, detail=str(e))

    def get_events(
        self,
        start: Optional[datetime] = None,
        end: Optional[datetime] = None
    ) -> List[Event]:
        """Get events with optional date range filtering"""
        query = self.db.query(Event)
        
        if start:
            query = query.filter(
                or_(
                    Event.start_time >= start,
                    Event.is_recurring == True
                )
            )
        
        if end:
            query = query.filter(
                Event.start_time <= end
            )
        
        return query.order_by(Event.start_time).all()

    def get_event_by_id(self, event_id: int) -> Optional[Event]:
        return self.db.query(Event).filter(Event.id == event_id).first()

    def update_event(self, event_id: int, event_update: EventCreate) -> Event:
        try:
            # Log raw input data
            logger.info(f"Update Event Request - ID: {event_id}")
            logger.info(f"Raw Event Update Data: {event_update}")
            
            # Validate input data types
            logger.debug(f"Input Data Types:")
            logger.debug(f"event_id: {type(event_id)}, value: {event_id}")
            logger.debug(f"name: {type(event_update.name)}, value: {event_update.name}")
            logger.debug(f"start_time: {type(event_update.start_time)}, value: {event_update.start_time}")
            logger.debug(f"duration: {type(event_update.duration)}, value: {event_update.duration}")
            logger.debug(f"is_recurring: {type(event_update.is_recurring)}, value: {event_update.is_recurring}")
            logger.debug(f"recurring_days: {type(event_update.recurring_days)}, value: {event_update.recurring_days}")
            
            # Retrieve existing event
            db_event = self.get_event_by_id(event_id)
            if not db_event:
                logger.error(f"Event with ID {event_id} not found")
                raise HTTPException(
                    status_code=404,
                    detail="Event not found"
                )
            
            # Log existing event details for comparison
            logger.debug(f"Existing Event Details: {db_event}")
            
            # Check for conflicts excluding the current event
            conflict_details = self._get_conflict_details(event_update, exclude_id=event_id)
            if conflict_details:
                logger.warning(f"Event update blocked due to conflicts: {conflict_details}")
                raise HTTPException(
                    status_code=400,
                    detail=f"Invalid event details: {conflict_details}"
                )
            
            # Convert event_update to dictionary for more robust updating
            update_dict = event_update.model_dump()
            
            # Explicitly handle recurring_days
            if 'recurring_days' in update_dict and update_dict['recurring_days']:
                # Ensure days are uppercase and sorted
                update_dict['recurring_days'] = ','.join(
                    sorted(str(day).upper() for day in update_dict['recurring_days'])
                )
            
            # Update event attributes
            for key, value in update_dict.items():
                try:
                    # Log each attribute update
                    logger.debug(f"Updating attribute {key}: {value}")
                    setattr(db_event, key, value)
                except Exception as attr_error:
                    logger.error(f"Error setting attribute {key}: {attr_error}")
                    raise HTTPException(
                        status_code=500,
                        detail=f"Error updating event attribute {key}: {str(attr_error)}"
                    )
            
            # Commit and refresh
            try:
                self.db.commit()
                self.db.refresh(db_event)
                
                # Log successful update
                logger.info(f"Event updated successfully: {db_event}")
                logger.debug(f"Updated Event Details: {db_event.__dict__}")
            except Exception as commit_error:
                self.db.rollback()
                logger.error(f"Database commit error: {commit_error}")
                raise HTTPException(
                    status_code=500,
                    detail=f"Database error during event update: {str(commit_error)}"
                )
            
            return db_event
        
        except Exception as e:
            # Catch-all error logging with full traceback
            logger.error(f"Unexpected error updating event: {str(e)}", exc_info=True)
            if not isinstance(e, HTTPException):
                raise HTTPException(
                    status_code=500,
                    detail=f"Unexpected error updating event: {str(e)}"
                )
            raise

    def delete_event(self, event_id: int) -> bool:
        db_event = self.get_event_by_id(event_id)
        if not db_event:
            raise HTTPException(
                status_code=404,
                detail="Event not found"
            )
        
        self.db.delete(db_event)
        self.db.commit()
        return True

    def get_current_events(self, current_time: datetime) -> List[Event]:
        """
        Retrieve events happening at the current time.
        
        Args:
            current_time: UTC datetime object
            
        Returns:
            List of events currently in progress
        """
        try:
            # Ensure current_time is UTC and naive (no timezone info)
            if current_time.tzinfo:
                current_time = current_time.replace(tzinfo=None)
            
            # Get current day code
            current_day_code = current_time.strftime('%a')[:2].upper()
            logger.debug(f"Searching for current events at {current_time} (Day: {current_day_code})")
            
            # Build the query using SQLAlchemy expressions
            non_recurring = and_(
                ~Event.is_recurring,
                Event.start_time <= current_time,
                func.datetime(
                    Event.start_time, 
                    '+' + func.cast(Event.duration, String) + ' minutes'
                ) >= current_time
            )
            
            # For recurring events, compare time components
            current_time_str = current_time.strftime('%H:%M:%S')
            
            recurring = and_(
                Event.is_recurring,
                Event.recurring_days.like(f'%{current_day_code}%'),
                func.time(Event.start_time) <= current_time_str,
                func.time(
                    func.datetime(
                        Event.start_time,
                        '+' + func.cast(Event.duration, String) + ' minutes'
                    )
                ) >= current_time_str
            )
            
            # Combine queries
            query = self.db.query(Event).filter(or_(non_recurring, recurring))
            
            # Execute query
            events = query.all()
            logger.info(f"Found {len(events)} current events")
            
            return events
            
        except Exception as e:
            logger.error(f"Error retrieving current events: {str(e)}", exc_info=True)
            raise HTTPException(
                status_code=500,
                detail=f"Error retrieving current events: {str(e)}"
            )

    def _get_conflict_details(self, event: EventCreate, exclude_id: Optional[int] = None) -> Optional[str]:
        """
        Check for event time conflicts using precise time overlap detection.
        
        Conflict is defined as:
        1. Events have overlapping time ranges
        2. Events are on the same day (for non-recurring events)
        3. Recurring events have additional conflict rules
        """
        try:
            # Calculate event end time
            event_end = event.start_time + timedelta(minutes=event.duration)
            
            # Log detailed conflict detection info
            logger.debug(f"Checking conflicts for event: {event}")
            logger.debug(f"Event start: {event.start_time}, Event end: {event_end}")
            
            # Base query for potential conflicts
            query = self.db.query(Event).filter(
                and_(
                    # Exclude the current event if an ID is provided
                    Event.id != (exclude_id or -1),
                    
                    # Check for time overlaps with more precise conditions
                    or_(
                        # Scenario 1: New event completely contains an existing event
                        and_(
                            event.start_time <= Event.start_time,
                            event_end >= Event.start_time + timedelta(minutes=Event.duration)
                        ),
                        # Scenario 2: Existing event completely contains the new event
                        and_(
                            Event.start_time <= event.start_time,
                            Event.start_time + timedelta(minutes=Event.duration) >= event_end
                        ),
                        # Scenario 3: Partial overlap at the start of the event
                        and_(
                            Event.start_time < event_end,
                            Event.start_time + timedelta(minutes=Event.duration) > event.start_time
                        )
                    ),
                    
                    # Additional conditions for non-recurring events
                    or_(
                        # For non-recurring events, check same day
                        and_(
                            ~Event.is_recurring,
                            ~event.is_recurring,
                            Event.start_time.date() == event.start_time.date()
                        ),
                        # For recurring events, check day of week
                        and_(
                            Event.is_recurring,
                            event.is_recurring,
                            # Check if recurring days overlap
                            or_(
                                # If both have recurring days
                                and_(
                                    Event.recurring_days is not None,
                                    event.recurring_days is not None,
                                    # Check for common days
                                    or_(
                                        *[
                                            Event.recurring_days.like(f'%{day}%') 
                                            for day in event.recurring_days or []
                                        ]
                                    )
                                ),
                                # If no specific recurring days, check general recurrence
                                Event.is_recurring
                            )
                        )
                    )
                )
            )
            
            # Log the raw SQL query for debugging
            logger.debug(f"Conflict detection query: {query.statement}")
            
            # Execute query and check for conflicts
            conflicts = query.all()
            
            # Log conflict details
            logger.debug(f"Number of conflicts found: {len(conflicts)}")
            
            if not conflicts:
                return None
            
            # Generate conflict details
            conflict_msgs = []
            for conflict in conflicts:
                conflict_end = conflict.start_time + timedelta(minutes=conflict.duration)
                conflict_msgs.append(
                    f"Conflict with event '{conflict.name}' "
                    f"(from {conflict.start_time.strftime('%Y-%m-%d %H:%M')} "
                    f"to {conflict_end.strftime('%Y-%m-%d %H:%M')}, "
                    f"Recurring: {conflict.is_recurring})"
                )
                
                # Additional logging for each conflict
                logger.debug(f"Conflict details: {conflict_msgs[-1]}")
            
            return "; ".join(conflict_msgs)
            
        except Exception as e:
            logger.error(f"Error checking conflicts: {str(e)}", exc_info=True)
            return None

File: app/src/app/services/event.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, throwError, of, switchMap } from 'rxjs';
import { tap, catchError } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { Event, EventCreate, EventList } from '../models/event.model';
import { EventValidationService } from './event-validation.service';
import moment from 'moment';

// Custom error class for event conflicts
export class EventConflictError extends Error {
  constructor(public conflictDetails: string) {
    super('Event not scheduled due to conflicts.');
    this.name = 'EventConflictError';
  }
}

// Interface to match backend event structure
export interface EventDetails {
  id: number;
  name: string;
  start_time: string;
  duration: number;
  is_recurring: boolean;
  recurring_days?: string | null;
}

@Injectable({
  providedIn: 'root'
})
export class EventService {
  private apiUrl = environment.apiUrl;
  private events: Event[] = [];

  constructor(
    private http: HttpClient,
    private eventValidation: EventValidationService
  ) {}

  getEvents(start?: Date, end?: Date): Observable<EventList> {
    console.log('Fetching Events from API', {
      startTime: start ? moment(start).format() : undefined,
      endTime: end ? moment(end).format() : undefined,
      timestamp: new Date().toISOString()
    });

    let params = new HttpParams();
    if (start) {
      params = params.set('start', moment(start).format());
    }
    if (end) {
      params = params.set('end', moment(end).format());
    }

    return this.http.get<EventList>(`${this.apiUrl}/events`, { params }).pipe(
      tap(eventList => {
        this.events = eventList.events; // Cache events for conflict checking
        console.log('Service Retrieved Events Successfully', {
          totalEvents: eventList.events.length,
          events: eventList.events,
          timestamp: new Date().toISOString()
        });

        // Log detailed event properties
        eventList.events.forEach(event => {
          console.log('Event Details:', {
            id: event.id,
            name: event.name,
            start_time: event.start_time,
            duration: event.duration,
            is_recurring: event.is_recurring,
            recurring_days: event.recurring_days
          });
        });
      }),
      catchError(error => {
        console.error('Service Event Retrieval Error', {
          error: error,
          timestamp: new Date().toISOString()
        });
        return throwError(() => new Error('Failed to fetch events'));
      })
    );
  }

  getEvent(id: number | string): Observable<Event> {
    console.log('Service Fetching Event from API', {
      eventId: id,
      timestamp: moment().format()
    });

    return this.http.get<Event>(`${this.apiUrl}/events/${id}`).pipe(
      tap(event => {
        console.log('Service Event Retrieved Successfully', {
          eventId: event.id,
          eventName: event.name,
          timestamp: moment().format()
        });
      }),
      catchError(error => {
        console.error('Service Error Fetching Event', {
          error: error,
          eventId: id,
          timestamp: moment().format()
        });
        return throwError(() => new Error('Failed to fetch event'));
      })
    );
  }

  getEventById(eventId: string): Observable<EventDetails> {
    return this.http.get<EventDetails>(`${this.apiUrl}/events/${eventId}`).pipe(
      tap(event => {
        console.log('Retrieved Event Details', {
          eventId,
          event,
          timestamp: new Date().toISOString()
        });
      }),
      catchError(error => {
        console.error('Error retrieving event details', {
          eventId,
          error,
          timestamp: new Date().toISOString()
        });
        return throwError(() => new Error('Failed to retrieve event details'));
      })
    );
  }

  createEvent(event: EventCreate): Observable<Event> {
    console.log('Service Creating Event', {
      eventData: event,
      timestamp: new Date().toISOString()
    });

    // Log detailed event properties
    console.log('Event Creation Details:', {
      name: event.name,
      start_time: event.start_time,
      duration: event.duration,
      is_recurring: event.is_recurring,
      recurring_days: event.recurring_days
    });

    // Validate start_time using moment
    if (!moment(event.start_time).isValid()) {
      return throwError(() => new Error('Invalid start time provided for event'));
    }

    // Format the event data using moment
    const formattedEvent: EventCreate = {
      ...event,
      start_time: moment(event.start_time).format('YYYY-MM-DDTHH:mm:ss.SSSZ')
    };

    // First check for conflicts
    return this.getEvents().pipe(
      switchMap(eventList => {
        const conflict = this.eventValidation.checkEventConflicts(formattedEvent, eventList.events);
        if (conflict) {
          return throwError(() => new EventConflictError(conflict));
        }
        
        // If no conflicts, create the event
        return this.http.post<Event>(`${this.apiUrl}/events`, formattedEvent);
      }),
      tap(createdEvent => {
        this.events.push(createdEvent); // Update cached events
        console.log('Service Event Created Successfully', {
          event: createdEvent,
          timestamp: new Date().toISOString()
        });
      }),
      catchError(error => {
        console.error('Service Event Creation Error', {
          error: error,
          eventData: event,
          timestamp: new Date().toISOString()
        });
        return throwError(() => error);
      })
    );
  }

  updateEvent(id: number | string, event: EventCreate): Observable<Event> {
    console.log('Service Updating Event', {
      eventId: id,
      event: event,
      timestamp: moment().format()
    });

    // Validate start_time using moment
    if (!moment(event.start_time).isValid()) {
      return throwError(() => new Error('Invalid start time provided for event'));
    }

    // Format the event data using moment
    const formattedEvent: EventCreate = {
      ...event,
      id,  // Include ID explicitly
      start_time: moment(event.start_time).format('YYYY-MM-DDTHH:mm:ss.SSSZ')
    };

    // First check for conflicts
    return this.getEvents().pipe(
      switchMap(eventList => {
        const conflict = this.eventValidation.checkEventConflicts(
          formattedEvent,  // Now includes ID
          eventList.events
        );
        if (conflict) {
          return throwError(() => new EventConflictError(conflict));
        }
        
        // If no conflicts, update the event
        return this.http.put<Event>(`${this.apiUrl}/events/${id}`, formattedEvent);
      }),
      tap(updatedEvent => {
        // Update cached events
        const index = this.events.findIndex(e => e.id === id);
        if (index !== -1) {
          this.events[index] = updatedEvent;
        }
        console.log('Service Event Updated Successfully', {
          event: updatedEvent,
          timestamp: moment().format()
        });
      }),
      catchError(error => {
        console.error('Service Error Updating Event', {
          error: error,
          eventId: id,
          eventData: event,
          timestamp: moment().format()
        });
        return throwError(() => error);
      })
    );
  }

  updateEventTime(eventId: string, newStartTime: string, newDuration: number): Observable<any> {
    // First, get the full event details
    return this.getEventById(eventId).pipe(
      switchMap(existingEvent => {
        // Prepare the full update payload
        const updatePayload: EventDetails = {
          id: existingEvent.id,
          name: existingEvent.name,
          start_time: newStartTime,
          duration: newDuration,
          is_recurring: existingEvent.is_recurring,
          // Convert recurring_days to comma-separated string or null
          recurring_days: existingEvent.recurring_days 
            ? (Array.isArray(existingEvent.recurring_days) 
                ? existingEvent.recurring_days.join(',') 
                : existingEvent.recurring_days)
            : null
        };

        console.log('Event Update Payload', {
          payload: updatePayload,
          timestamp: new Date().toISOString()
        });

        return this.http.put<any>(`${this.apiUrl}/events/${existingEvent.id}`, updatePayload).pipe(
          tap(response => {
            console.log('Event time updated successfully', {
              eventId,
              newStartTime,
              newDuration,
              response,
              timestamp: new Date().toISOString()
            });
          }),
          catchError(error => {
            console.error('Detailed Error updating event time', {
              eventId,
              payload: updatePayload,
              errorStatus: error.status,
              errorBody: error.error,
              errorMessage: error.message,
              timestamp: new Date().toISOString()
            });
            
            // If there's a specific error message from the backend, use it
            const errorMessage = error.error?.message || 
              error.error?.detail || 
              'Failed to update event time';
            
            return throwError(() => new Error(errorMessage));
          })
        );
      }),
      catchError(error => {
        console.error('Error in event update process', {
          eventId,
          error,
          timestamp: new Date().toISOString()
        });
        return throwError(() => error);
      })
    );
  }

  deleteEvent(id: number | string): Observable<void> {
    console.log('Service Deleting Event', {
      eventId: id,
      timestamp: moment().format()
    });

    return this.http.delete<void>(`${this.apiUrl}/events/${id}`).pipe(
      tap(() => {
        // Update cached events
        this.events = this.events.filter(e => e.id !== id);
        console.log('Service Event Deleted Successfully', {
          eventId: id,
          timestamp: moment().format()
        });
      }),
      catchError(error => {
        console.error('Error Deleting Event', {
          error: error,
          eventId: id,
          timestamp: moment().format()
        });
        return throwError(() => new Error('Failed to delete event'));
      })
    );
  }
}
